internal/consensus/state.go

```diff
@@ -1506,6 +1506,21 @@ func (cs *State) defaultDoPrevote(ctx context.Context, height int64, round int32
 		}
 	}
 
+	stateMachineValidBlock, err := cs.blockExec.ProcessProposal(ctx, cs.ProposalBlock)
+	if err != nil {
+		panic(fmt.Sprintf(
+			"state machine returned an error (%v) when calling ProcessProposal", err,
+		))
+	}
+
+	// Vote nil if application invalidated the block
+	if !stateMachineValidBlock {
+		// Consensus says we must vote nil
+		logger.Error("prevote step: consensus deems this block to be mustVoteNil", "err", err)
```

Comment:
Can we clean up the comments here i.e. it's the application that his deemed the block invalid not consensus (I know this was part of the earlier commit). Also I'm unsure whether it makes sense to log an error. Logged errors are telling the node operator that they should be doing something. There isn't really anything they can do in this case. I would simply log this as either `Info` or `Debug`

Reply:
Actually, after reflecting on this, I left the log at error level. If Application rejects blocks, according to the spec, someone is byzantine, so I think operators should be aware of it. I clarified the error message